File: test/router.test.ts
import { describe, expect, it } from 'vitest';
import { chooseGroupedModel, chooseModel, orderedCandidates } from '../src/latency/router.js';
import { allGroupModelIds, selectedGroupModelIds } from '../src/model-groups.js';

describe('group-name router', () => {
  it('routes to group when requestedModel matches group name', () => {
    const groups = { fast: ['a', 'b'], balanced: ['c'] };
    expect(chooseModel(groups, 'fast')).toEqual({ modelId: 'a', reason: 'model-group' });
  });

  it('falls back to defaultGroup when requestedModel is not a group', () => {
    const groups = { fast: ['a', 'b'], balanced: ['c'] };
    expect(orderedCandidates(groups, 'auto')).toEqual(['a', 'b']);
  });

  it('falls back to first group when no request', () => {
    const groups = { default: ['z', 'a'] };
    expect(chooseModel(groups, undefined)).toEqual({ modelId: 'z', reason: 'fallback-order' });
  });

  it('returns group order as candidates', () => {
    const groups = { default: ['a', 'b', 'c'] };
    expect(orderedCandidates(groups)).toEqual(['a', 'b', 'c']);
  });

  it('routes group aliases within the configured group only', () => {
    const groups = { fast: ['b'], balanced: ['a'], capable: ['c'] };
    expect(chooseGroupedModel(groups, 'slr/fast')).toEqual({ modelId: 'b', reason: 'model-group' });
    expect(orderedCandidates(groups, 'haiku')).toEqual(['b']);
  });

  it('returns empty array when all groups are empty', () => {
    expect(orderedCandidates({ fast: [], balanced: [], capable: [] }, 'opus')).toEqual([]);
  });
});

describe('configurable groups', () => {
  it('routes to a custom group by name', () => {
    const groups = { coding: ['model-a', 'model-b'], chat: ['model-c'] };
    expect(chooseGroupedModel(groups, 'coding'))
      .toEqual({ modelId: 'model-a', reason: 'model-group' });
  });

  it('returns all models in group as ordered candidates', () => {
    const groups = { coding: ['model-a', 'model-b', 'model-c'] };
    expect(orderedCandidates(groups, 'coding'))
      .toEqual(['model-a', 'model-b', 'model-c']);
  });

  it('falls back to defaultGroup when requested model is not a group name', () => {
    const groups = { coding: ['model-a', 'model-b'], default: ['model-c', 'model-d'] };
    expect(orderedCandidates(groups, 'unknown-model', 'default'))
      .toEqual(['model-c', 'model-d']);
  });

  it('falls back to first group when defaultGroup is not set', () => {
    const groups = { coding: ['model-a', 'model-b'], chat: ['model-c'] };
    expect(orderedCandidates(groups, 'unknown-model'))
      .toEqual(['model-a', 'model-b']);
  });

  it('uses defaultGroup for generic model names like auto', () => {
    const groups = { fast: ['model-a'], default: ['model-b', 'model-c'] };
    expect(orderedCandidates(groups, 'auto', 'default'))
      .toEqual(['model-b', 'model-c']);
  });

  it('ignores slr/ prefix on group names', () => {
    const groups = { coding: ['model-a', 'model-b'] };
    expect(chooseGroupedModel(groups, 'slr/coding'))
      .toEqual({ modelId: 'model-a', reason: 'model-group' });
  });

  it('supports legacy aliases (haiku, sonnet, opus)', () => {
    const groups = { fast: ['model-a'], balanced: ['model-b'], capable: ['model-c'] };
    expect(chooseGroupedModel(groups, 'haiku'))
      .toEqual({ modelId: 'model-a', reason: 'model-group' });
    expect(chooseGroupedModel(groups, 'sonnet'))
      .toEqual({ modelId: 'model-b', reason: 'model-group' });
    expect(chooseGroupedModel(groups, 'opus'))
      .toEqual({ modelId: 'model-c', reason: 'model-group' });
  });

  it('derives master list from all groups', () => {
    const groups = { a: ['x', 'y'], b: ['y', 'z'] };
    expect(allGroupModelIds(groups)).toEqual(['x', 'y', 'z']);
  });

  it('returns undefined from selectedGroupModelIds for unknown group', () => {
    const groups = { coding: ['model-a'] };
    expect(selectedGroupModelIds(groups, 'unknown')).toBeUndefined();
  });
});

File: router_test.go
package sleepyrouter

import (
	"testing"
)

// Tests mirrored from test/router.test.ts

func TestRouter_RoutesToGroup(t *testing.T) {
	groups := ModelGroups{"fast": {"a", "b"}, "balanced": {"c"}}
	choice, err := ChooseModel(groups, "fast")
	if err != nil {
		t.Fatal(err)
	}
	if choice.ModelID != "a" || choice.Reason != RouteModelGroup {
		t.Fatalf("choice: %v", choice)
	}
}

func TestRouter_FallbackToDefaultGroup(t *testing.T) {
	groups := ModelGroups{"fast": {"a", "b"}, "balanced": {"c"}}
	ids := OrderedCandidates(groups, "auto", "", "fast", "balanced")
	if len(ids) != 2 || ids[0] != "a" || ids[1] != "b" {
		t.Fatalf("candidates: %v", ids)
	}
}

func TestRouter_NoRequestFallsBack(t *testing.T) {
	groups := ModelGroups{"default": {"z", "a"}}
	choice, err := ChooseModel(groups, "")
	if err != nil {
		t.Fatal(err)
	}
	if choice.ModelID != "z" || choice.Reason != RouteFallbackOrder {
		t.Fatalf("choice: %v", choice)
	}
}

func TestRouter_GroupOrderAsCandidates(t *testing.T) {
	groups := ModelGroups{"default": {"a", "b", "c"}}
	ids := OrderedCandidates(groups, "", "")
	if len(ids) != 3 || ids[0] != "a" || ids[1] != "b" || ids[2] != "c" {
		t.Fatalf("candidates: %v", ids)
	}
}

func TestRouter_GroupAliases(t *testing.T) {
	groups := ModelGroups{"fast": {"b"}, "balanced": {"a"}, "capable": {"c"}}
	choice, err := ChooseGroupedModel(groups, "sleepyrouter/fast", "")
	if err != nil {
		t.Fatal(err)
	}
	if choice.ModelID != "b" || choice.Reason != RouteModelGroup {
		t.Fatalf("choice: %v", choice)
	}
	ids := OrderedCandidates(groups, "haiku", "")
	if len(ids) != 1 || ids[0] != "b" {
		t.Fatalf("haiku candidates: %v", ids)
	}
}

func TestRouter_EmptyGroups(t *testing.T) {
	ids := OrderedCandidates(ModelGroups{"fast": {}, "balanced": {}, "capable": {}}, "opus", "")
	if len(ids) != 0 {
		t.Fatalf("expected empty, got %v", ids)
	}
}

func TestRouter_CustomGroupByName(t *testing.T) {
	groups := ModelGroups{"coding": {"model-a", "model-b"}, "chat": {"model-c"}}
	choice, err := ChooseGroupedModel(groups, "coding", "")
	if err != nil {
		t.Fatal(err)
	}
	if choice.ModelID != "model-a" || choice.Reason != RouteModelGroup {
		t.Fatalf("choice: %v", choice)
	}
}

func TestRouter_AllModelsInGroupAsCandidates(t *testing.T) {
	groups := ModelGroups{"coding": {"model-a", "model-b", "model-c"}}
	ids := OrderedCandidates(groups, "coding", "")
	if len(ids) != 3 || ids[0] != "model-a" || ids[1] != "model-b" || ids[2] != "model-c" {
		t.Fatalf("candidates: %v", ids)
	}
}

func TestRouter_FallbackToDefaultGroupWhenNotAGroup(t *testing.T) {
	groups := ModelGroups{"coding": {"model-a", "model-b"}, "default": {"model-c", "model-d"}}
	ids := OrderedCandidates(groups, "unknown-model", "default")
	if len(ids) != 2 || ids[0] != "model-c" || ids[1] != "model-d" {
		t.Fatalf("candidates: %v", ids)
	}
}

func TestRouter_FallbackToFirstGroupWhenDefaultNotSet(t *testing.T) {
	groups := ModelGroups{"coding": {"model-a", "model-b"}, "chat": {"model-c"}}
	ids := OrderedCandidates(groups, "unknown-model", "", "coding", "chat")
	if len(ids) != 2 || ids[0] != "model-a" || ids[1] != "model-b" {
		t.Fatalf("candidates: %v", ids)
	}
}

func TestRouter_DefaultGroupForAuto(t *testing.T) {
	groups := ModelGroups{"fast": {"model-a"}, "default": {"model-b", "model-c"}}
	ids := OrderedCandidates(groups, "auto", "default")
	if len(ids) != 2 || ids[0] != "model-b" || ids[1] != "model-c" {
		t.Fatalf("candidates: %v", ids)
	}
}

func TestRouter_IgnoreSleepyRouterPrefix(t *testing.T) {
	groups := ModelGroups{"coding": {"model-a", "model-b"}}
	choice, err := ChooseGroupedModel(groups, "sleepyrouter/coding", "")
	if err != nil {
		t.Fatal(err)
	}
	if choice.ModelID != "model-a" || choice.Reason != RouteModelGroup {
		t.Fatalf("choice: %v", choice)
	}
}

func TestRouter_LegacyAliases(t *testing.T) {
	groups := ModelGroups{"fast": {"model-a"}, "balanced": {"model-b"}, "capable": {"model-c"}}
	c1, err := ChooseGroupedModel(groups, "haiku", "")
	if err != nil {
		t.Fatal(err)
	}
	if c1.ModelID != "model-a" || c1.Reason != RouteModelGroup {
		t.Fatalf("haiku: %v", c1)
	}
	c2, err := ChooseGroupedModel(groups, "sonnet", "")
	if err != nil {
		t.Fatal(err)
	}
	if c2.ModelID != "model-b" || c2.Reason != RouteModelGroup {
		t.Fatalf("sonnet: %v", c2)
	}
	c3, err := ChooseGroupedModel(groups, "opus", "")
	if err != nil {
		t.Fatal(err)
	}
	if c3.ModelID != "model-c" || c3.Reason != RouteModelGroup {
		t.Fatalf("opus: %v", c3)
	}
}

func TestRouter_AllGroupModelIDs(t *testing.T) {
	groups := ModelGroups{"a": {"x", "y"}, "b": {"y", "z"}}
	ids := AllGroupModelIDs(groups)
	if len(ids) != 3 || ids[0] != "x" || ids[1] != "y" || ids[2] != "z" {
		t.Fatalf("all group ids: %v", ids)
	}
}

func TestRouter_SelectedGroupModelIDsUnknown(t *testing.T) {
	groups := ModelGroups{"coding": {"model-a"}}
	ids := SelectedGroupModelIDs(groups, "unknown")
	if ids != nil {
		t.Fatalf("expected nil, got %v", ids)
	}
}
